platform/darwin/src/MGLStyleValue_Private.h

```diff
@@ -115,162 +117,14 @@ class MGLStyleValueTransformer {
-                                    NSCAssert(mbglStopValue.isConstant(), @"Categorical stop keys must be strings, booleans, or integers");
-                                } else if ([key compare:@(0)] == NSOrderedDescending ||
-                                           [key compare:@(0)] == NSOrderedSame ||
-                                           [key compare:@(0)] == NSOrderedAscending) {
-                                    stops[mbgl::style::CategoricalValue((int64_t)[key integerValue])] = mbglStopValue.asConstant();
-                                }
-                            }
-                        }];
-                        outerStops[zoomKey.floatValue] = stops;
-                    }];
-                    mbgl::style::CompositeFunction<MBGLType> compositeFunction {
-                        compositeStyleFunction.attributeName.UTF8String,
-                        mbgl::style::CompositeCategoricalStops<MBGLType> { outerStops }
-                    };
-                    if (compositeStyleFunction.defaultValue) {
-                        NSCAssert([compositeStyleFunction.defaultValue isKindOfClass:[MGLStyleConstantValue class]], @"Default value must be constant");
-                        MBGLType mbglValue;
-                        id mglValue = [(MGLStyleConstantValue<ObjCType> *)compositeStyleFunction.defaultValue rawValue];
-                        getMBGLValue(mglValue, mbglValue);
-                        compositeFunction.defaultValue = mbglValue;
-                    }
-                    return compositeFunction;
-                }
-                    break;
-                default:
-                    [NSException raise:NSInvalidArgumentException
-                                format:@"A composite function must use exponential, interval, or categorical stops."];
-                    return {};
-            }
-            return {};
-        } else if ([value isMemberOfClass:[MGLStyleFunction class]]) {
-            MGLStyleFunction<ObjCType> *styleFunction = (MGLStyleFunction<ObjCType> *)value;
-            return toMBGLExponentialCameraFunction(styleFunction);
-        } else if (value) {
-            [NSException raise:@"MGLAbstractClassException" format:
-             @"The style value %@ cannot be applied to the style. "
-             @"Make sure the style value was created as a member of a concrete subclass of MGLStyleValue.",
-             NSStringFromClass([value class])];
-            return {};
-        } else {
-            return {};
+        // TODO: check that it's an MGLStyleFunction before casting.
+        NSObject *rawValue = toRawStyleSpecValue((MGLStyleFunction<ObjCType> *) value);
+        mbgl::style::conversion::Result<mbgl::style::DataDrivenPropertyValue<MBGLType>> result = mbgl::style::conversion::convert<mbgl::style::DataDrivenPropertyValue<MBGLType>>(rawValue);
+        if (!result) {
+            // raise an exception?
```

Comment:
In Objective-C, it’s extremely rare (read: kludge) to `@try`-`@catch` an NSException. (`NSAssert()` and `NSCAssert()` both raise NSExceptions under the hood.) In fact, on iOS, an uncaught exception will crash the application. On macOS, an uncaught exception will spew to the console and skip to the next iteration of the run loop, which may either be graceful or lead to maddeningly undebuggable downstream misbehavior.

If you don’t want the application to crash or misbehave, warn with `NSLog()` or whatever C++ warning facility you have at your disposal.